### DISCUSSION TITLE
Pretty Tracebacks with links in PyCharm

### DISCUSSION THREAD

**Discussion by lucmos** [Textualize/rich#875] [Q&A]
Hello,

I'd like to use the pretty tracebacks in PyCharm.

Currently, my traceback looks like this:
![image](https://user-images.githubusercontent.com/11019190/103484811-de7b1f00-4df1-11eb-9413-e51211e3f6b5.png)

It's uglier than the rich's tracebacks, but it contains "links" that let the editor jump directly to the file and line in question.

Using rich, the path is still there:
![image](https://user-images.githubusercontent.com/11019190/103484840-12eedb00-4df2-11eb-8d74-050e2538fff9.png)
But, it is *not* clickable anymore.

Do you know how can I make the links work?

---

There is a regexp in PyCharm that catches the pattern

```
print('File <path>, line <int>')
```

The line number is optional, e.g. this opens the file manager correctly
![image](https://user-images.githubusercontent.com/11019190/103485031-a4ab1800-4df3-11eb-830e-2e710c73d05d.png)

Unfortunately at the moment, I can't find a way to change the regexp or change the formatting in the rich output. Is there?

---

Adding the `File` keyword and some quotes is enough to make it work (correctly, even if the underlying is wrong)
![image](https://user-images.githubusercontent.com/11019190/103485130-8abe0500-4df4-11eb-9019-e5531a10e972.png)

**Comment by willmcgugan** [comment]
Some terminals turn FILE:LINE in to a clickable link, but will generally not know how to handle Python tracebacks. PyCharm is obviously aware of Python tracebacks and knows how to extract path and link.

Unfortunately if I make it work in PyCharm, it would break clickable links in the terminal. I could be persuaded to add a PyCharm mode for tracebacks to get the best of both worlds.

BTW you might want to check the [note in the docs about PyCharm](https://rich.readthedocs.io/en/latest/introduction.html), you may be able to get even prettier tracebacks.

**Comment by lucmos** [comment]
Thanks a lot for the tip! 


I think I'll open an issue on the PyCharm side, there is a problem with their regexp:

This is correctly recognized:

```python
print('FILE:LINE')
```
This is not:
```python
print('| FILE:LINE')
```

**Reply by LarsHill** [reply]
Is there any update? Did you create a PyCharm issue? If yes, I think it makes sense to link it here.

Btw other special characters, e.g. 
```python
print('\ FILE:LINE')
``` 
work fine. Only `|` seems to break the regex which is a pitty since rich uses `|` to create the nice frame.

**Reply by lucmos** [reply]
Yes, I opened an issue on the Pycharm side some time ago

https://youtrack.jetbrains.com/issue/PY-46305

**Comment by viktor-yunenko** [comment]
Thanks to lucmos' comment I was able to do a monkey-patch that can be placed eg in django's settings.py

```python
import rich.box
import rich.traceback
from rich.panel import Panel


panel_init_original = Panel.__init__


def panel_init_fixed(self, *args, **kwargs):
    is_traceback_panel = "Traceback" in kwargs.get("title", "")
    if is_traceback_panel:
        box_pycharm_compatible = rich.box.MINIMAL_DOUBLE_HEAD
        kwargs["box"] = box_pycharm_compatible

        for arg in args:
            # handle rich code that passes Box as a positional arg
            if isinstance(arg, rich.box.Box):
                args = list(args)
                args.remove(arg)
                args = tuple(args)

    panel_init_original(self, *args, **kwargs)


Panel.__init__ = panel_init_fixed

rich.traceback.install(with=300)
```
